File: warmup/day2/group-anagrams.js
// 13 September, 2025
// I did not how to sort the string in cpp so I solved this with JavaScript :)

/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    let m = new Map();
    for (let str of strs) {
        let sorted = str.split('').sort().join('');
        if (m.has(sorted)) {
            let prev = m.get(sorted);
            prev.push(str);
            m.delete(sorted);
            m.set(sorted, prev);
        } else {
            m.set(sorted, [str])
        }
    }
    return Array.from(m.values());
};

File: warmup/day2/group-anagrams.cpp
// 13 September, 2025
//
// sort(v.begin(), v.end()) - learned that sort works for strings as well like ints

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> m;
        
        for (string str : strs) {
            string sorted = str;
            sort(sorted.begin(), sorted.end());
            m[sorted].push_back(str);
        }

        vector<vector<string>> res;
        for (auto entry : m) {
            res.push_back(entry.second);
        }
        return res;
    }
};

